test/common/http/conn_manager_impl_test.cc

```diff
@@ -887,6 +887,201 @@ TEST_F(HttpConnectionManagerImplTest, RouteShouldUseSantizedPath) {
+
+  std::shared_ptr<Router::MockRoute> foo_bar_baz_route =
+      std::make_shared<NiceMock<Router::MockRoute>>();
+
+  std::shared_ptr<Router::MockRoute> foo_bar_route =
+      std::make_shared<NiceMock<Router::MockRoute>>();
+  EXPECT_CALL(foo_bar_route->route_entry_, clusterName()).WillOnce(ReturnRef(foo_bar_cluster_name));
+
+  std::shared_ptr<Router::MockRoute> foo_route = std::make_shared<NiceMock<Router::MockRoute>>();
+
+  std::shared_ptr<Router::MockRoute> default_route =
+      std::make_shared<NiceMock<Router::MockRoute>>();
+  EXPECT_CALL(default_route->route_entry_, clusterName())
+      .Times(2)
+      .WillRepeatedly(ReturnRef(default_cluster_name));
+
+  using ::testing::InSequence;
+  {
+    InSequence seq;
+    EXPECT_CALL(*route_config_provider_.route_config_, route(_, _, _, _))
+        .WillOnce(Return(default_route));
+
+    // This filter iterates through all possible route matches and choose the last matched route
+    EXPECT_CALL(*decoder_filters_[0], decodeHeaders(_, true))
+        .WillOnce(InvokeWithoutArgs([&]() -> FilterHeadersStatus {
+          EXPECT_EQ(default_route, decoder_filters_[0]->callbacks_->route());
+          EXPECT_EQ(default_route->routeEntry(),
+                    decoder_filters_[0]->callbacks_->streamInfo().routeEntry());
+          EXPECT_EQ(default_cluster->info(), decoder_filters_[0]->callbacks_->clusterInfo());
+
+          // Not clearing cached route returns cached route and doesn't invoke cb.
+          Router::RouteConstSharedPtr route = decoder_filters_[0]->callbacks_->route(
+              [](Router::RouteConstSharedPtr, Router::RouteEvalStatus) -> Router::RouteMatchStatus {
+                ADD_FAILURE() << "When route cache is not cleared CB should not be invoked";
+                return Router::RouteMatchStatus::Accept;
+              });
+          EXPECT_EQ(default_route, route);
+
+          int ctr = 0;
+          const Router::RouteCallback& cb =
+              [&](Router::RouteConstSharedPtr route,
+                  Router::RouteEvalStatus route_eval_status) -> Router::RouteMatchStatus {
+            EXPECT_LE(ctr, 3);
+            if (ctr == 0) {
+              ++ctr;
+              EXPECT_EQ(foo_bar_baz_route, route);
+              EXPECT_EQ(route_eval_status, Router::RouteEvalStatus::HasMoreRoutes);
+              return Router::RouteMatchStatus::Continue;
+            }
+
+            if (ctr == 1) {
+              ++ctr;
+              EXPECT_EQ(foo_bar_route, route);
+              EXPECT_EQ(route_eval_status, Router::RouteEvalStatus::HasMoreRoutes);
+              return Router::RouteMatchStatus::Continue;
+            }
+
+            if (ctr == 2) {
+              ++ctr;
+              EXPECT_EQ(foo_route, route);
+              EXPECT_EQ(route_eval_status, Router::RouteEvalStatus::HasMoreRoutes);
+              return Router::RouteMatchStatus::Continue;
+            }
+
+            if (ctr == 3) {
+              ctr++;
```

Comment:
nit: just for the sake of consistency id say use the same unary ++ operator as the other branches (`++ctr` instead of `ctr++`).